File: mobile/src/main/java/rs/readahead/washington/mobile/views/fragment/BlankFormsListFragment.java
package rs.readahead.washington.mobile.views.fragment;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import org.hzontal.shared_ui.bottomsheet.CustomBottomSheetFragment;
import org.hzontal.shared_ui.utils.DialogUtils;
import org.javarosa.core.model.FormDef;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.LinearLayoutManager;
import rs.readahead.washington.mobile.R;
import rs.readahead.washington.mobile.data.sharedpref.Preferences;
import rs.readahead.washington.mobile.databinding.FragmentBlankFormsListBinding;
import rs.readahead.washington.mobile.domain.entity.collect.CollectForm;
import rs.readahead.washington.mobile.domain.entity.collect.ListFormResult;
import rs.readahead.washington.mobile.javarosa.FormUtils;
import rs.readahead.washington.mobile.mvp.contract.ICollectBlankFormListPresenterContract;
import rs.readahead.washington.mobile.mvp.presenter.CollectBlankFormListPresenter;
import rs.readahead.washington.mobile.util.DialogsUtil;
import rs.readahead.washington.mobile.util.ViewUtil;
import rs.readahead.washington.mobile.views.activity.CollectMainActivity;
import rs.readahead.washington.mobile.views.adapters.CollectDownloadedFormInstanceRecycleViewAdapter;
import rs.readahead.washington.mobile.views.custom.TopSpaceItemDecoration;
import timber.log.Timber;

public class BlankFormsListFragment extends FormListFragment implements
        ICollectBlankFormListPresenterContract.IView,
        BlankFormListListener {
    private CollectBlankFormListPresenter presenter;
    private AlertDialog alertDialog;
    private boolean silentFormUpdates = false;
    private CollectDownloadedFormInstanceRecycleViewAdapter adapter;

    private FragmentBlankFormsListBinding binding;

    public static BlankFormsListFragment newInstance() {
        return new BlankFormsListFragment();
    }

    @Override
    public Type getFormListType() {
        return Type.BLANK;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        adapter = new CollectDownloadedFormInstanceRecycleViewAdapter(this);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentBlankFormsListBinding.inflate(inflater, container, false);
        View rootView = binding.getRoot();

        createPresenter();

        binding.downloadedForms.setLayoutManager(new LinearLayoutManager(requireContext()));
        binding.downloadedForms.addItemDecoration(new TopSpaceItemDecoration(ViewUtil.getDpInPixels(requireContext(), 16)));
        binding.downloadedForms.setAdapter(adapter);

        return rootView;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        // listBlankForms();
        refreshBlankForms();
    }

    @Override
    public void onStart() {
        super.onStart();
    }

    @Override
    public void onDestroy() {
        destroyPresenter();
        hideAlertDialog();
        super.onDestroy();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    @Override
    public void showBlankFormRefreshLoading() {
        if (alertDialog != null) return;
        if (getActivity() != null) {
            ((CollectMainActivity) getActivity()).hideFab();
        }
        if (!silentFormUpdates) {
            alertDialog = DialogsUtil.showCollectRefreshProgressDialog(getContext(), (dialog, which) -> presenter.userCancel());
        }
    }

    private void showBlankFormDownloadingDialog(int progressText) {
        if (alertDialog != null) return;
        if (getActivity() != null) {
            ((CollectMainActivity) getActivity()).hideFab();
        }
        alertDialog = DialogsUtil.showFormUpdatingDialog(getContext(), (dialog, which) -> presenter.userCancel(), progressText);
    }

    @Override
    public void hideBlankFormRefreshLoading() {
        Preferences.setLastCollectRefresh(System.currentTimeMillis());
        if (silentFormUpdates) {
            silentFormUpdates = false;
        }
        hideAlertDialog();
    }

    @Override
    public void onDownloadBlankFormDefSuccess(CollectForm collectForm) {
    }

    @Override
    public void onDownloadBlankFormDefStart() {
        showBlankFormDownloadingDialog(R.string.collect_dialog_text_download_progress);
    }

    @Override
    public void onDownloadBlankFormDefEnd() {
        hideAlertDialog();
        Toast.makeText(getActivity(), R.string.collect_toast_download_completed, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onUpdateBlankFormDefStart() {
        showBlankFormDownloadingDialog(R.string.collect_blank_dialog_expl_updating_form_definitions);
    }

    @Override
    public void onUpdateBlankFormDefEnd() {
        hideAlertDialog();
        Toast.makeText(getActivity(), R.string.collect_blank_toast_form_definition_updated, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onBlankFormDefRemoved(String formName) {
        DialogUtils.showBottomMessage(requireActivity(), String.format("%s has been deleted.", formName), false); // todo: string
    }

    @Override
    public void onBlankFormDefRemoveError(Throwable error) {
        // todo: show error
    }

    @Override
    public void onUpdateBlankFormDefSuccess(CollectForm collectForm, FormDef formDef) {
        updateDownloadedFormList();
    }

    @Override
    public void onUserCancel() {
        hideAlertDialog();
        Toast.makeText(getActivity(), R.string.collect_blank_toast_refresh_canceled, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onFormDefError(Throwable error) {
        String errorMessage = FormUtils.getFormDefErrorMessage(requireContext(), error);
        Toast.makeText(getActivity(), errorMessage, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onBlankFormsListResult(ListFormResult listFormResult) {
        adapter.setForms(listFormResult.getForms());
    }

    @Override
    public void onBlankFormsListError(Throwable throwable) {
        Timber.d(throwable, getClass().getName());
    }

    @Override
    public void onNoConnectionAvailable() {
        if (!silentFormUpdates) {
            Toast.makeText(getActivity(), R.string.collect_blank_toast_not_connected, Toast.LENGTH_SHORT).show();
        }
    }

    public void showDeleteBottomSheet(final CollectForm form) {
        final CustomBottomSheetFragment bottomSheet = CustomBottomSheetFragment.Companion.with(requireActivity().getSupportFragmentManager())
                .page(R.layout.form_fragment_delete_btm_sheet)
                .cancellable(true);

        bottomSheet.holder(new DeleteBottomSheetHolder(), holder -> {
            holder.title.setText(form.getForm().getName());
            holder.title2.setText(String.format("Delete \"%s\"?", form.getForm().getName())); // todo: string
            holder.delete.setOnClickListener(v -> {
                holder.firstStep.setVisibility(View.GONE);
                holder.secondStep.setVisibility(View.VISIBLE);
            });
            holder.yes.setOnClickListener(v -> {
                presenter.removeBlankFormDef(form);
                bottomSheet.dismiss();
            });
            holder.no.setOnClickListener(v -> {
                // todo: remove this
                presenter.downloadBlankFormDef(form);
                bottomSheet.dismiss();
            });
        }).transparentBackground().launch();
    }

    private void updateDownloadedFormList() {
    }

    public void listBlankForms() {
        if (presenter != null) {
            presenter.listBlankForms();
        }
    }

    public void refreshBlankForms() {
        if (presenter != null) {
            presenter.refreshBlankForms();
        }
    }

    private void hideAlertDialog() {
        if (alertDialog != null) {
            alertDialog.dismiss();
            alertDialog = null;
        }
        if (getActivity() != null) {
            ((CollectMainActivity) getActivity()).showFab();
        }
    }

    private void createPresenter() {
        if (presenter == null) {
            presenter = new CollectBlankFormListPresenter(this);
        }
    }

    private void destroyPresenter() {
        if (presenter != null) {
            presenter.destroy();
            presenter = null;
        }
    }

    static class DeleteBottomSheetHolder extends CustomBottomSheetFragment.PageHolder {
        public ViewGroup firstStep;
        public ViewGroup secondStep;
        public TextView title;
        public TextView delete;
        public TextView title2;
        public TextView no;
        public TextView yes;

        @Override
        public void bindView(@NonNull View view) {
            firstStep = view.findViewById(R.id.first_step);
            secondStep = view.findViewById(R.id.second_step);
            title = view.findViewById(R.id.title);
            delete = view.findViewById(R.id.delete);
            title2 = view.findViewById(R.id.title2);
            no = view.findViewById(R.id.no);
            yes = view.findViewById(R.id.yes);
        }
    }
}

File: mobile/src/main/java/org/horizontal/tella/mobile/views/fragment/forms/BlankFormsListFragment.kt
package org.horizontal.tella.mobile.views.fragment.forms

import android.Manifest
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.LinearLayout
import androidx.appcompat.app.AlertDialog
import androidx.core.app.ActivityCompat
import androidx.core.content.res.ResourcesCompat
import androidx.fragment.app.viewModels
import dagger.hilt.android.AndroidEntryPoint
import org.horizontal.tella.mobile.MyApplication
import org.horizontal.tella.mobile.R
import org.horizontal.tella.mobile.data.sharedpref.Preferences
import org.horizontal.tella.mobile.databinding.BlankCollectFormRowBinding
import org.horizontal.tella.mobile.databinding.FragmentBlankFormsListBinding
import org.horizontal.tella.mobile.domain.entity.collect.CollectForm
import org.horizontal.tella.mobile.domain.entity.collect.ListFormResult
import org.horizontal.tella.mobile.javarosa.FormUtils
import org.horizontal.tella.mobile.util.C
import org.horizontal.tella.mobile.util.DialogsUtil
import org.horizontal.tella.mobile.views.activity.CollectFormEntryActivity
import org.horizontal.tella.mobile.views.activity.MainActivity
import org.horizontal.tella.mobile.views.base_ui.BaseBindingFragment
import org.hzontal.shared_ui.bottomsheet.BottomSheetUtils
import org.hzontal.shared_ui.bottomsheet.BottomSheetUtils.ActionSeleceted
import org.hzontal.shared_ui.utils.DialogUtils
import org.javarosa.core.model.FormDef
import permissions.dispatcher.NeedsPermission
import timber.log.Timber

@AndroidEntryPoint
class BlankFormsListFragment :
    BaseBindingFragment<FragmentBlankFormsListBinding>(FragmentBlankFormsListBinding::inflate),
    FormListInterface {

    private val model: SharedFormsViewModel by viewModels()

    private var availableForms: MutableList<CollectForm>? = null
    private var downloadedForms: MutableList<CollectForm>? = null
    private var alertDialog: AlertDialog? = null
    private var noUpdatedForms = 0
    private var silentFormUpdates = false
    override fun getFormListType(): FormListInterface.Type {
        return FormListInterface.Type.BLANK
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        availableForms = ArrayList()
        downloadedForms = ArrayList()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initObservers()
        /* if (!Preferences.isJavarosa3Upgraded()) {
             model.showFab.postValue(false)
             showJavarosa2UpgradeSheet()
         } else {*/
        listBlankForms()

        //}
    }

    override fun onDestroy() {
        hideAlertDialog()
        super.onDestroy()
    }

    private fun showBlankFormDownloadingDialog(progressText: Int) {
        if (alertDialog != null) return
        if (activity != null) {
            //model.showFab.postValue(false)
            showFab(false)
        }
        alertDialog = DialogsUtil.showFormUpdatingDialog(
            context,
            { _: DialogInterface?, _: Int -> model.userCancel() }, progressText
        )
    }

    private fun initObservers() {

        binding.fab.setOnClickListener {
            refreshBlankForms()
        }

        model.onToggleFavoriteSuccess.observe(viewLifecycleOwner) { updatedForm ->
            updatedForm?.let { updateFormList(it) } // Update only changed item
        }
        model.onError.observe(viewLifecycleOwner) { error ->
            Timber.d(error, javaClass.name)
        }
        model.onGetBlankFormDefSuccess.observe(viewLifecycleOwner) { result ->
            result.let {
                startCreateFormControllerPresenter(it.form, it.formDef)
            }
        }
        model.onCreateFormController.observe(viewLifecycleOwner) {
            if (Preferences.isAnonymousMode()) {
                startCollectFormEntryActivity() // no need to check for permissions, as location won't be turned on
            } else {
                if (!hasLocationPermissions(baseActivity)) {
                    requestLocationPermissions()
                } else {
                    startCollectFormEntryActivity() // no need to check for permissions, as location won't be turned on
                }
            }
        }
        model.showBlankFormRefreshLoading.observe(
            viewLifecycleOwner
        ) { show: Boolean? ->
            if (!show!!) {
                Preferences.setLastCollectRefresh(
                    System.currentTimeMillis()
                )
                if (silentFormUpdates) {
                    silentFormUpdates = false
                }
                hideAlertDialog()
            } else {
                if (alertDialog != null) return@observe
                if (activity != null) {
                    //model.showFab.postValue(false)
                    showFab(false)
                }
                if (!silentFormUpdates) {
                    alertDialog = DialogsUtil.showCollectRefreshProgressDialog(
                        context
                    ) { _: DialogInterface?, _: Int -> model.userCancel() }
                }
            }
        }
        model.onDownloadBlankFormDefSuccess.observe(
            viewLifecycleOwner
        ) { form: CollectForm? -> updateForm(form!!) }
        model.onDownloadBlankFormDefStart.observe(
            viewLifecycleOwner
        ) { show: Boolean? ->
            if (show == true) {
                showBlankFormDownloadingDialog(R.string.collect_dialog_text_download_progress)
            } else {
                hideAlertDialog()
                DialogUtils.showBottomMessage(
                    baseActivity,
                    getString(R.string.collect_toast_download_completed),
                    false
                )
            }
        }
        model.onUpdateBlankFormDefStart.observe(
            viewLifecycleOwner
        ) { show: Boolean ->
            if (show) {
                showBlankFormDownloadingDialog(R.string.collect_blank_dialog_expl_updating_form_definitions)
            } else {
                hideAlertDialog()
                DialogUtils.showBottomMessage(
                    baseActivity,
                    getString(R.string.collect_blank_toast_form_definition_updated),
                    false
                )
            }
        }
        model.onBlankFormDefRemoved.observe(
            viewLifecycleOwner
        ) { updateFormViews() }
        model.onUpdateBlankFormDefSuccess.observe(
            viewLifecycleOwner
        ) { (first, second): Pair<CollectForm?, FormDef?> ->
            onUpdateBlankFormDefSuccess(
                first,
                second
            )
        }
        model.onUserCancel.observe(viewLifecycleOwner) {
            hideAlertDialog()
            DialogUtils.showBottomMessage(
                baseActivity,
                getString(R.string.collect_blank_toast_refresh_canceled),
                false
            )
        }
        model.onFormDefError.observe(
            viewLifecycleOwner
        ) { error: Throwable? ->
            onFormDefError(
                error!!
            )
        }
        model.onFormCacheCleared.observe(viewLifecycleOwner) {
            refreshBlankForms()
            //model.showFab.postValue(true)
            showFab(true)
        }
        model.onBlankFormsListResult.observe(
            viewLifecycleOwner
        ) { listFormResult: ListFormResult? ->
            onBlankFormsListResult(
                listFormResult!!
            )
        }
        model.onNoConnectionAvailable.observe(
            viewLifecycleOwner
        ) {
            if (!silentFormUpdates) {
                DialogUtils.showBottomMessage(
                    baseActivity,
                    getString(R.string.collect_blank_toast_not_connected),
                    true
                )
            }
        }
        /*model.showFab.observe(viewLifecycleOwner, { show ->
            binding.fab.isVisible = show
        })*/
    }

    private fun onUpdateBlankFormDefSuccess(collectForm: CollectForm?, formDef: FormDef?) {
        noUpdatedForms -= 1
        showBanner()
        updateDownloadedFormList()
    }

    private fun onFormDefError(error: Throwable) {
        val errorMessage = FormUtils.getFormDefErrorMessage(requireContext(), error)
        DialogUtils.showBottomMessage(
            baseActivity,
            errorMessage,
            true
        )
    }

    private fun onBlankFormsListResult(listFormResult: ListFormResult) {
        updateFormLists(listFormResult)
        showBanner()
        updateFormViews()
        if (context != null && MyApplication.isConnectedToInternet(context) && checkIfDayHasPassed()) {
            silentFormUpdates = true
            refreshBlankForms()
        }
    }

    private fun updateFormLists(listFormResult: ListFormResult) {
        noUpdatedForms = 0
        binding.blankFormView.visibility = View.VISIBLE
        downloadedForms!!.clear()
        availableForms!!.clear()
        binding.blankFormsInfo.visibility =
            if (listFormResult.forms.isEmpty()) View.VISIBLE else View.GONE
        for (form in listFormResult.forms) {
            if (form.isDownloaded) {
                downloadedForms!!.add(form)
                if (form.isUpdated) {
                    noUpdatedForms += 1
                }
            } else {
                availableForms!!.add(form)
            }
        }
        // todo: make this multiply errors friendly
        if (!silentFormUpdates) {
            for (error in listFormResult.errors) {
                DialogUtils.showBottomMessage(
                    baseActivity,
                    String.format(
                        "%s %s",
                        getString(R.string.collect_blank_toast_fail_updating_form_list),
                        error.serverName
                    ),
                    true
                )
                Timber.d(error.exception, javaClass.name)
            }
        }
    }

    private fun updateDownloadedFormList() {
        updateFormViews()
    }

    private fun listBlankForms() {
        model.listBlankForms()
    }

    private fun refreshBlankForms() {
        model.refreshBlankForms()
    }

    private fun updateForm(form: CollectForm) {
        availableForms!!.remove(form)
        downloadedForms!!.add(form)
        updateFormViews()
    }

    private fun setViewsVisibility() {
        binding.downloadedFormsTitle.visibility =
            if (downloadedForms!!.size > 0) View.VISIBLE else View.GONE
        binding.downloadedForms.visibility =
            if (downloadedForms!!.size > 0) View.VISIBLE else View.GONE
        binding.avaivableFormsTitle.visibility =
            if (availableForms!!.size > 0) View.VISIBLE else View.GONE
        binding.blankForms.visibility =
            if (availableForms!!.size > 0) View.VISIBLE else View.GONE
    }

    private fun updateFormViews() {
        binding.downloadedForms.removeAllViews()
        binding.blankForms.removeAllViews()
        createCollectFormViews(availableForms!!, binding.blankForms)
        createCollectFormViews(downloadedForms!!, binding.downloadedForms)
        setViewsVisibility()
    }

    private fun hideAlertDialog() {
        if (alertDialog != null) {
            alertDialog!!.dismiss()
            alertDialog = null
        }
        /*if (activity != null) {
            model.showFab.postValue(true)
            showFab(true)
        }*/
    }

    private fun createCollectFormViews(forms: List<CollectForm>, listView: LinearLayout) {
        for (form in forms) {
            val view = getCollectFormItem(form)
            listView.addView(view, forms.indexOf(form))
        }
    }

    private fun getCollectFormItem(collectForm: CollectForm?): View {
        val itemBinding =
            BlankCollectFormRowBinding.inflate(LayoutInflater.from(context), binding.forms, false)
        val row = itemBinding.formRow
        val name = itemBinding.name
        val organization = itemBinding.organization
        val dlOpenButton = itemBinding.dlOpenButton
        val pinnedIcon = itemBinding.favoritesButton
        val rowLayout = itemBinding.rowLayout
        val updateButton = itemBinding.laterButton
        if (collectForm != null) {
            name.text = collectForm.form.name
            organization.text = collectForm.serverName
            if (collectForm.isDownloaded) {
                dlOpenButton.setImageDrawable(
                    ResourcesCompat.getDrawable(
                        resources,
                        R.drawable.ic_more,
                        null
                    )
                )
                dlOpenButton.contentDescription =
                    getString(R.string.collect_blank_action_desc_more_options)
                dlOpenButton.setOnClickListener {
                    showDownloadedMenu(
                        collectForm
                    )
                }
                rowLayout.setOnClickListener {
                    model.getBlankFormDef(
                        collectForm
                    )
                }
                pinnedIcon.setOnClickListener {
                    model.toggleFavorite(collectForm)
                }
                if (collectForm.isUpdated) {
                    pinnedIcon.visibility = View.VISIBLE
                    updateButton.visibility = View.VISIBLE
                    updateButton.setOnClickListener {
                        if (MyApplication.isConnectedToInternet(requireContext())) {
                            model.updateBlankFormDef(collectForm)
                        } else {
                            DialogUtils.showBottomMessage(
                                baseActivity,
                                getString(R.string.collect_blank_toast_not_connected),
                                true
                            )
                        }
                    }
                } else {
                    updateButton.visibility = View.GONE
                }
            } else {
                pinnedIcon.visibility = View.GONE
                dlOpenButton.setImageDrawable(
                    ResourcesCompat.getDrawable(
                        resources,
                        R.drawable.ic_download,
                        null
                    )
                )
                dlOpenButton.contentDescription =
                    getString(R.string.collect_blank_action_download_form)
                dlOpenButton.setOnClickListener {
                    if (MyApplication.isConnectedToInternet(requireContext())) {
                        model.downloadBlankFormDef(collectForm)
                    } else {
                        DialogUtils.showBottomMessage(
                            baseActivity,
                            getString(R.string.collect_blank_toast_not_connected),
                            true
                        )
                    }
                }
            }
            if (collectForm.isPinned) {
                pinnedIcon.setImageDrawable(
                    ResourcesCompat.getDrawable(
                        resources,
                        R.drawable.star_filled_24dp,
                        null
                    )
                )
                pinnedIcon.contentDescription = getString(R.string.action_unfavorite)
            } else {
                pinnedIcon.setImageDrawable(
                    ResourcesCompat.getDrawable(
                        resources,
                        R.drawable.star_border_24dp,
                        null
                    )
                )
                pinnedIcon.contentDescription = getString(R.string.action_favorite)
            }
        }
        return itemBinding.root
    }

    private fun updateFormList(updatedForm: CollectForm) {
        val indexInDownloaded = downloadedForms?.indexOfFirst { it.id == updatedForm.id }
        val indexInAvailable = availableForms?.indexOfFirst { it.id == updatedForm.id }

        if (indexInDownloaded != null && indexInDownloaded != -1) {
            downloadedForms?.set(indexInDownloaded, updatedForm)
        } else if (indexInAvailable != null && indexInAvailable != -1) {
            availableForms?.set(indexInAvailable, updatedForm)
        }

        updateFormViews() // Refresh UI properly
    }

    private fun hasLocationPermissions(context: Context): Boolean {
        return ActivityCompat.checkSelfPermission(
            context,
            Manifest.permission.ACCESS_FINE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(
            context,
            Manifest.permission.ACCESS_COARSE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED
    }

    private fun requestLocationPermissions() {
        baseActivity.maybeChangeTemporaryTimeout()
        val permissions = arrayOf(
            Manifest.permission.ACCESS_COARSE_LOCATION,
            Manifest.permission.ACCESS_FINE_LOCATION
        )
        ActivityCompat.requestPermissions(
            //1
            baseActivity,
            //2
            permissions,
            //3
            LOCATION_REQUEST_CODE
        )
    }

    private fun startCreateFormControllerPresenter(form: CollectForm, formDef: FormDef) {
        model.createFormController(form, formDef)
    }

    @NeedsPermission(Manifest.permission.ACCESS_FINE_LOCATION)
    fun startCollectFormEntryActivity() {
        startActivity(Intent(activity, CollectFormEntryActivity::class.java))
    }

    private fun showDownloadedMenu(collectForm: CollectForm) {
        BottomSheetUtils.showEditDeleteMenuSheet(
            requireActivity().supportFragmentManager,
            collectForm.form.name,
            requireContext().getString(R.string.Collect_Action_FillForm),
            requireContext().getString(R.string.action_delete),
            object : ActionSeleceted {
                override fun accept(action: BottomSheetUtils.Action) {
                    if (action === BottomSheetUtils.Action.EDIT) {
                        model.getBlankFormDef(collectForm)
                    }
                    if (action === BottomSheetUtils.Action.DELETE) {
                        downloadedForms!!.remove(collectForm)
                        model.removeBlankFormDef(collectForm)
                        updateFormViews()
                    }
                }
            },
            requireContext().getString(R.string.Collect_RemoveForm_SheetTitle),
            String.format(
                requireContext().resources.getString(R.string.Collect_Subtitle_RemoveForm),
                collectForm.form.name
            ),
            requireContext().getString(R.string.action_remove),
            requireContext().getString(R.string.action_cancel)
        )
    }

    private fun checkIfDayHasPassed(): Boolean {
        val lastRefresh = Preferences.getLastCollectRefresh()
        return System.currentTimeMillis() - lastRefresh > C.DAY
    }

    private fun showBanner() {
        if (noUpdatedForms > 0) {
            binding.banner.visibility = View.VISIBLE
        } else {
            binding.banner.visibility = View.GONE
        }
    }

    private fun showFab(show: Boolean) {
        if (show) {
            binding.fab.visibility = View.VISIBLE
        } else {
            binding.fab.visibility = View.GONE
        }
    }


    private fun upgradeJavarosa2() {
        try {
            DialogUtils.showBottomMessage(
                baseActivity,
                requireActivity().getString(R.string.Javarosa_Upgrade_Toast),
                false
            )
            model.deleteCachedForms()
        } catch (t: Throwable) {
            Timber.d(t)
        }
    }

    private fun goHome() {
        if (activity == null) return
        (requireActivity() as MainActivity).selectHome()
    }

    companion object {
        fun newInstance(): BlankFormsListFragment {
            return BlankFormsListFragment()
        }
    }
}